src/ride/track.c

```diff
@@ -4921,15 +4922,446 @@ void game_command_remove_track(int *eax, int *ebx, int *ecx, int *edx, int *esi,
+}
+
+money32 loc_6CDB35(uint16 x, uint8 flags, uint8 direction, uint16 y, uint8 rideIndex, uint8 mode, uint16 unknown, uint16 z, rct_map_element *mapElement) {
+	uint8 bl;
+
+	if ((flags & GAME_COMMAND_FLAG_APPLY) == 0) {
+		return loc_6CDDFD();
+	}
+
+
+	if (mode == 1) {
+		return loc_6CDDA2(x, y, rideIndex, mapElement);
+	}
+
+	if (mode == 0) {
+		uint8 edx = get_edx(x, y);
+
+		// 0x6cdcff
+		mapElement->properties.track.maze_entry &= ~(1 << edx);
+
+		if (direction == 4) {
+			return loc_6CDDA2(x, y, rideIndex, mapElement);
+		}
+
+		// 0x6cdd12
+		edx = RCT2_GLOBAL(0x993ce9 + (direction + edx), uint8);
+		// TODO: wth is edx?
+		mapElement->properties.track.maze_entry &= ~(1 << edx);
+
+		// SAVE edx + esi
+
+		// TODO: btr     [esi+5], dx
+		// (esi+5 = mapElement->properties.track.maze_entry)
+
+		uint32 temp_edx = RCT2_GLOBAL(0x993cfc + edx, uint8);
+		if ((temp_edx & 0xFF) == 0xFF) {
+			return loc_6CDDA2(x, y, rideIndex, mapElement);
+		}
+
+		uint16 xRounded = floor2(x, 32);
+		uint16 yRounded = floor2(y, 32);
+
+		uint32 ebx = direction;
+		xRounded -= TileDirectionDelta[direction].x;
+		yRounded -= TileDirectionDelta[direction].y;
+
+		uint8 bh = rideIndex;
+		bl = (uint8) ((z >> 3) & 0xFF);
+
+		rct_map_element *temp_esi = map_get_first_element_at(xRounded / 32, yRounded / 32);
+
+		// 0x6cdd71:
+		do {
+			if (map_element_get_type(temp_esi) != MAP_ELEMENT_TYPE_TRACK) {
+				continue;
+			}
+
+			if (temp_esi->base_height != bl) {
+				continue;
+			}
+
+			if (temp_esi->properties.track.ride_index != rideIndex) {
+				continue;
+			}
+
+			if (temp_esi->properties.track.type != 0x65) {
+				continue;
+			}
+
+			// loc_6CDD9B;
+			temp_esi->properties.track.maze_entry &= ~(1 << temp_edx);
+
+			return loc_6CDDA2(x, y, rideIndex, mapElement);
+		} while (!map_element_is_last_for_tile(temp_esi++));
+
+		// 6cdd92
+		mapElement->properties.track.maze_entry |= (1 << edx);
+		return loc_6CDDA2(x, y, rideIndex, mapElement);
+	}
+
+
+	// 0x6cdb53
+	if (direction == 4) {
+		return loc_6CDDA2(x, y, rideIndex, mapElement);
+	}
+
+	// 0x6cdb69
+
+	uint16 ax = x - TileDirectionDelta[direction].x / 2;
+	uint16 cx = y - TileDirectionDelta[direction].y / 2;
+
+	bl = (z >> 3) & 0xFF;
+
+	mapElement = map_get_first_element_at(ax / 32, cx / 32);
+
+	// eax = incremented X;
+	map_invalidate_tile_full(floor2(ax, 32), floor2(cx, 32));
+
+	// eax = ax;
+	// ecx = cx;
+	// ebx = [bl: bl, bh: rideIndex];
+
+	do {
+		if (map_element_get_type(mapElement) != MAP_ELEMENT_TYPE_TRACK) {
+			continue;
+		}
+
+		if (mapElement->base_height != bl) {
+			continue;
+		}
+
+		if (mapElement->properties.track.ride_index != rideIndex) {
+			continue;
+		}
+
+		if (mapElement->properties.track.type != 0x65) {
+			continue;
+		}
+
+		return loc_6CDBE9(x, y, ax, cx, z, rideIndex, mapElement);
+		// loc_6CDBDB:
+	} while (!map_element_is_last_for_tile(mapElement++));
+
+	log_error("No surface found");
+	return loc_6CDDA2(x, y, rideIndex, mapElement);
```

Comment:
Returning MONEY32_UNDEFINED would be more appropriate than trying to perform actions on an invalid mapElement.